src/async/windows_event_loop.zig

```diff
@@ -0,0 +1,340 @@
+const bun = @import("root").bun;
+const Output = bun.Output;
+const JSC = bun.JSC;
+const uws = bun.uws;
+const Environment = bun.Environment;
+const std = @import("std");
+const uv = bun.windows.libuv;
+pub const Loop = uv.Loop;
+
+pub const KeepAlive = struct {
+    // handle.init zeroes the memory
+    handle: uv.uv_async_t = undefined,
+
+    status: Status = .inactive,
+
+    const log = Output.scoped(.KeepAlive, false);
+
+    const Status = enum { active, inactive, done };
+
+    pub inline fn isActive(this: KeepAlive) bool {
+        if (comptime Environment.allow_assert) {
+            if (this.status == .active) {
+                std.debug.assert(this.handle.isActive());
+            }
+        }
+        return this.status == .active;
+    }
+
+    /// Make calling ref() on this poll into a no-op.
+    pub fn disable(this: *KeepAlive) void {
+        if (this.status == .active) {
+            this.unref(JSC.VirtualMachine.get());
+            this.handle.close(null);
+        }
+
+        this.status = .done;
+    }
+
+    /// Only intended to be used from EventLoop.Pollable
+    pub fn deactivate(this: *KeepAlive, loop: *Loop) void {
+        _ = loop;
+        if (this.status != .active)
+            return;
+
+        this.status = .inactive;
+        this.handle.close(null);
+    }
+
+    /// Only intended to be used from EventLoop.Pollable
+    pub fn activate(this: *KeepAlive, loop: *Loop) void {
+        if (this.status != .active)
+            return;
+
+        this.status = .active;
+        this.handle.init(loop, null);
+    }
+
+    pub fn init() KeepAlive {
+        return .{};
+    }
+
+    /// Prevent a poll from keeping the process alive.
+    pub fn unref(this: *KeepAlive, vm: *JSC.VirtualMachine) void {
+        _ = vm;
+        if (this.status != .active)
+            return;
+        this.status = .inactive;
+        this.handle.unref();
+    }
+
+    /// From another thread, Prevent a poll from keeping the process alive.
+    pub fn unrefConcurrently(this: *KeepAlive, vm: *JSC.VirtualMachine) void {
+        _ = vm;
+        if (this.status != .active)
+            return;
+        this.status = .inactive;
+        this.handle.unref();
```

Comment:
Is this called from a different thread than the one that calls uv_run()? If so, that's not safe, you can only call uv_async_send() from another thread, not uv_ref() or uv_unref().

Reply:
thank you